core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -504,15 +511,63 @@ public Object visitGetAllCommand(InvocationContext ctx, GetAllCommand command) t
                }
             }
          }
-         Map<Object, Object> values = (Map<Object, Object>) invokeNextInterceptor(ctx, command);
-         return values;
+         return invokeNextInterceptor(ctx, command);
+      }
+   }
+
+   protected void wrapInternalCacheEntry(InternalCacheEntry ice, InvocationContext ctx, Object key,
+                                         boolean isWrite, FlagAffectedCommand command) {
+      if (!ctx.replaceValue(key, ice))  {
+         if (isWrite)
+            entryFactory.wrapEntryForPut(ctx, key, ice, false, command, true);
+         else
+            ctx.putLookedUpEntry(key, ice);
       }
    }
 
    /**
     * @return Whether a remote get is needed to obtain the previous values of the affected entries.
     */
-   protected abstract boolean needValuesFromPreviousOwners(InvocationContext ctx, WriteCommand command);
+   protected abstract boolean writeNeedsRemoteValue(InvocationContext ctx, WriteCommand command, Object key);
+
+   protected boolean valueIsMissing(CacheEntry entry) {
+      return entry == null || (entry.isNull() && !entry.isRemoved() && !entry.skipLookup());
+   }
+
+   protected abstract boolean previousValueIsRemote(Object key);
 
    protected abstract void remoteGetBeforeWrite(InvocationContext ctx, WriteCommand command, Object key) throws Throwable;
+
+   protected boolean readNeedsRemoteGet(InvocationContext ctx, AbstractDataCommand command) {
+      if (command.hasFlag(Flag.CACHE_MODE_LOCAL) || command.hasFlag(Flag.SKIP_REMOTE_LOOKUP)) {
+         return false;
+      }
+      Object key = command.getKey();
+      ConsistentHash ch = stateTransferManager.getCacheTopology().getReadConsistentHash();
+      boolean shouldFetchFromRemote = ctx.isOriginLocal() && !isValueAvailableLocally(ch, key);
+      if (!shouldFetchFromRemote && getLog().isTraceEnabled()) {
+         getLog().tracef("Not doing a remote get for key %s since entry is mapped to current node (%s) or is in L1. Owners are %s", toStr(key), rpcManager.getAddress(), ch.locateOwners(key));
+      }
+      return shouldFetchFromRemote;
+   }
+
+   protected boolean isValueAvailableLocally(ConsistentHash consistentHash, Object key) {
+      if (consistentHash.isKeyLocalToNode(rpcManager.getAddress(), key)) {
+         return true;
+      } else if (isL1Enabled) {
+         InternalCacheEntry entry = dataContainer.get(key);
+         return entry != null && entry.isL1Entry();
+      }
+      return false;
+   }
+
+   protected InternalCacheEntry fetchValueLocallyIfAvailable(ConsistentHash consistentHash, Object key) {
+      if (consistentHash.isKeyLocalToNode(rpcManager.getAddress(), key)) {
+         return dataContainer.get(key);
+      } else if (isL1Enabled) {
+         InternalCacheEntry entry = dataContainer.get(key);
+         return entry != null && entry.isL1Entry() ? entry : null;
```

Comment:
It can happen, if the consistent hash changed but we haven't yet removed the values that we no longer own: https://issues.jboss.org/browse/ISPN-4444